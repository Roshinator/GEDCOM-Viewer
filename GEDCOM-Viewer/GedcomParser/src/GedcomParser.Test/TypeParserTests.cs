using System.Collections.Generic;
using System.Linq;
using System.Text;
using ApprovalTests;
using ApprovalTests.Reporters;
using GedcomParser.Services;
using GedcomParser.Test.Services;
using Shouldly;
using Xunit;


namespace GedcomParser.Test
{
    [UseReporter(typeof(DiffReporter))]
    public class TypeParserTests
    {
        [Fact]
        public void CanParseKennedyFamily()
        {
            // Arrange
            const string fileName = "CustomSample.Kennedy.ged";
            var lines = ResourceHelper.GetLines(fileName);

            // Act
            var result = TypeParser.ParseTypeStructure(lines);

            // Assert
            result.Keys.Count.ShouldBe(36);
            result.Values.Sum(v => v.Count).ShouldBe(45);
            var output = GetAllTypes(fileName, result);
            Approvals.Verify(output);
        }

        [Fact]
        public void CanParseWindsorFamily()
        {
            // Arrange
            const string fileName = "CustomSample.Windsor.ged";
            var lines = ResourceHelper.GetLines(fileName);

            // Act
            var result = TypeParser.ParseTypeStructure(lines);

            // Assert
            result.Keys.Count.ShouldBe(17);
            result.Values.Sum(v => v.Count).ShouldBe(17);
            var output = GetAllTypes(fileName, result);
            Approvals.Verify(output);
        }

        [Fact]
        public void CanParseStefanFamily()
        {
            // Arrange
            const string fileName = "CustomSample.Stefan.ged";
            var lines = ResourceHelper.GetLines(fileName);

            // Act
            var result = TypeParser.ParseTypeStructure(lines);

            // Assert
            result.Keys.Count.ShouldBe(78);
            result.Values.Sum(v => v.Count).ShouldBe(176);
            var output = GetAllTypes(fileName, result);
            Approvals.Verify(output);
        }

        [Fact]
        public void CanParseGedcomStandardSample()
        {
            // Arrange
            const string fileName = "GedcomStandard.555SAMPLE.GED";
            var lines = ResourceHelper.GetLines(fileName);

            // Act
            var result = TypeParser.ParseTypeStructure(lines);

            // Assert
            result.Keys.Count.ShouldBe(48);
            result.Values.Sum(v => v.Count).ShouldBe(62);
            var output = GetAllTypes(fileName, result);
            Approvals.Verify(output);
        }

        [Fact]
        public void CanParseGedcomTortureSample()
        {
            // Arrange
            const string fileName = "GedcomTorture.TGC551.ged";
            var lines = ResourceHelper.GetLines(fileName);

            // Act
            var result = TypeParser.ParseTypeStructure(lines);

            // Assert
            result.Keys.Count.ShouldBe(129);
            result.Values.Sum(v => v.Count).ShouldBe(489);
            var output = GetAllTypes(fileName, result);
            Approvals.Verify(output);
        }

        private static string GetAllTypes(string fileName, IDictionary<string, List<string>> types)
        {
            var sb = new StringBuilder();
            sb.AppendLine("======================================================================");
            sb.AppendLine($"  ALL types detected in the file : '{fileName}'");
            sb.AppendLine("======================================================================");

            foreach (var type in types.Keys)
            {
                var subTypes = types[type];
                sb.AppendLine(type);
                foreach (var subType in subTypes)
                {
                    var indicator = types[subType].Any() ? "+" : "";
                    sb.AppendLine($"    {subType}{indicator}");
                }
            }

            return sb.ToString();
        }

    }
}