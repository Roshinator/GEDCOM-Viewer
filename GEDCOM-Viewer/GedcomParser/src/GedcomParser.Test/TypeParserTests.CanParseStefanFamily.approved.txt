======================================================================
  ALL types detected in the file : 'CustomSample.Stefan.ged'
======================================================================
_FREL
_GRP
    NAME+
    NOTE+
_MREL
_PLC
    _PLP
    ABBR+
    FORM
    NAME+
    NOTE+
    RIN
    TYPE
_PLP
_PRIM
_PRNT
_REL
_SIZE
_UID
ABBR
    CONC
ADDR
    CONT
AUTH
BAPM
    _PLC+
    DATE+
    NOTE+
    PLAC+
BIRT
    _PLC+
    DATE+
    NOTE+
    OBJE+
    PLAC+
    SOUR+
BURI
    _PLC+
    DATE+
    NOTE+
    PLAC+
    SOUR+
CAUS
CHAN
    DATE+
CHAR
CHIL
    _FREL
    _MREL
CHR
    _PLC+
    DATE+
    NOTE+
    OBJE+
    PLAC+
    SOUR+
CONC
CONF
    NOTE+
    PLAC+
CONT
CORP
DATA
    DATE+
    TEXT+
DATE
    TIME
DEAT
    _PLC+
    CAUS
    DATE+
    NOTE+
    OBJE+
    PLAC+
    SOUR+
DEST
DIV
    _PLC+
    DATE+
    NOTE+
    PLAC+
    SOUR+
DSCR
    _PLC+
    DATE+
    PLAC+
    SOUR+
EDUC
    DATE+
    PLAC+
EMAIL
EMIG
    _PLC+
    DATE+
    NOTE+
    PLAC+
    SOUR+
EVEN
    _PLC+
    DATE+
    NOTE+
    OBJE+
    PLAC+
    SOUR+
    TYPE
FACT
    _PLC+
    DATE+
    PLAC+
    SOUR+
FAM
    _REL
    CHAN+
    CHIL+
    DIV+
    EVEN+
    HUSB
    MARR+
    NOTE+
    SOUR+
    WIFE
FAMC
FAMS
FILE
FORM
GEDC
    FORM
    VERS
GIVN
HEAD
    CHAR
    DATE+
    DEST
    FILE
    GEDC+
    LANG
    SOUR+
    SUBM+
HUSB
IMMI
    _PLC+
    DATE+
    NOTE+
    PLAC+
    SOUR+
INDI
    _GRP+
    _UID
    BAPM+
    BIRT+
    BURI+
    CHAN+
    CHR+
    CONF+
    DEAT+
    DSCR+
    EDUC+
    EMIG+
    EVEN+
    FACT+
    FAMC
    FAMS
    IMMI+
    NAME+
    NCHI
    NMR
    NOTE+
    OBJE+
    OCCU+
    RESI+
    RIN
    SEX
    SOUR+
LANG
LATI
LONG
MAP
    LATI
    LONG
MARR
    _PLC+
    DATE+
    NOTE+
    PLAC+
    SOUR+
NAME
    GIVN
    NICK
    SOUR+
    SURN
NCHI
NICK
NMR
NOTE
    CONC
    CONT
OBJE
    _PRIM
    _PRNT
    _SIZE
    FILE
    FORM
    TITL
OCCU
    _PLC+
    DATE+
    PLAC+
    SOUR+
PAGE
PHON
PLAC
    MAP+
PUBL
QUAY
REPO
RESI
    _PLC+
    ADDR+
    DATE+
    NOTE+
    OBJE+
    PLAC+
    SOUR+
    TYPE
RIN
SEX
SOUR
    ABBR+
    AUTH
    CORP
    DATA+
    NOTE+
    PAGE
    PUBL
    QUAY
    REPO
    TEXT+
    TITL
    VERS
SUBM
    ADDR+
    EMAIL
    NAME+
    PHON
SURN
TEXT
    CONC
    CONT
TIME
TITL
TRLR
TYPE
VERS
WIFE
